substrate/frame/bounties/src/lib.rs

```diff
@@ -1008,57 +1551,279 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+				(curator.is_pending_or_failed(), beneficiary.is_pending_or_failed()),
+			_ => unreachable!(),
+		};
+		ensure!(process_curator || process_beneficiary, Error::<T, I>::UnexpectedStatus);
+
+		let bounty_account = Self::bounty_account_id(bounty_id, bounty.asset_kind.clone())?;
+		let (final_fee, payout) =
+			Self::calculate_curator_fee_and_payout(bounty_id, bounty.fee, bounty.value);
+
+		// Retry curator payout if needed
+		if process_curator {
+			let id = <T as pallet::Config<I>>::Paymaster::pay(
+				&bounty_account,
+				&curator_stash.0,
+				bounty.asset_kind.clone(),
+				final_fee,
+			)
+			.map_err(|_| Error::<T, I>::PayoutError)?;
+			curator_status = Some(PaymentState::Attempted { id });
+			Self::deposit_event(Event::<T, I>::Paid { index: bounty_id, payment_id: id });
+		}
+
+		// Retry beneficiary payout if needed
+		if process_beneficiary {
+			let id = <T as pallet::Config<I>>::Paymaster::pay(
+				&bounty_account,
+				&beneficiary.0,
+				bounty.asset_kind.clone(),
+				payout,
+			)
+			.map_err(|_| Error::<T, I>::PayoutError)?;
+			beneficiary_status = Some(PaymentState::Attempted { id });
+			Self::deposit_event(Event::<T, I>::Paid { index: bounty_id, payment_id: id });
+		}
+
+		// Both will always be `Some` if we are here
+		Ok((
+			curator_status.unwrap_or(PaymentState::Pending),
+			beneficiary_status.unwrap_or(PaymentState::Pending),
+		))
+	}
+
+	fn do_check_funding_payment_status(
+		bounty_id: BountyIndex,
+		bounty: &BountyOf<T, I>,
+		payment_status: PaymentState<PaymentIdOf<T, I>>,
+	) -> Result<PaymentState<PaymentIdOf<T, I>>, DispatchError> {
+		let payment_id = payment_status.get_attempt_id().ok_or(Error::<T, I>::UnexpectedStatus)?;
+
+		match <T as pallet::Config<I>>::Paymaster::check_payment(payment_id) {
+			PaymentStatus::Success => {
+				let err_amount = T::Currency::unreserve(&bounty.proposer, bounty.bond);
+				debug_assert!(err_amount.is_zero());
+				Self::deposit_event(Event::<T, I>::BountyBecameActive { index: bounty_id });
+				Ok(PaymentState::Succeeded)
+			},
+			PaymentStatus::InProgress => return Err(Error::<T, I>::FundingInconclusive.into()),
+			PaymentStatus::Unknown | PaymentStatus::Failure => {
+				Self::deposit_event(Event::<T, I>::PaymentFailed { index: bounty_id, payment_id });
+				return Ok(PaymentState::Failed)
+			},
+		}
+	}
+
+	fn do_check_refund_payment_status(
+		bounty_id: BountyIndex,
+		bounty: &BountyOf<T, I>,
+		payment_status: PaymentState<PaymentIdOf<T, I>>,
+		curator: Option<T::AccountId>,
+	) -> Result<PaymentState<PaymentIdOf<T, I>>, DispatchError> {
+		let payment_id = payment_status.get_attempt_id().ok_or(Error::<T, I>::UnexpectedStatus)?;
+
+		match <T as pallet::Config<I>>::Paymaster::check_payment(payment_id) {
+			PaymentStatus::Success => {
+				if let Some(curator) = curator {
+					// Cancelled by council, refund deposit of the working curator.
+					let err_amount = T::Currency::unreserve(&curator, bounty.curator_deposit);
```

Comment:
this is behaviour not obvious from the function name, I would add the doc for the function